### PR TITLE
fix(transparent-proxy): make iptables mode detection more defensive

### PR DESCRIPTION
**Pull request by bartsmykla**
### Checklist prior to review

- [x] [Link to relevant issue][1] as well as docs and UI issues
  - No relevant issues
- [x] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS
  - It shouldn't
- [x] Tests (Unit test, E2E tests, manual test on universal and k8s)
  - Manually tested:
    - minikube (kvm2) without `iptables-nft` - thank you, @michaelbeaumont
      ```sh
      Linux minikube 5.10.57 #1 SMP Tue Nov 7 06:51:54 UTC 2023 x86_64 GNU/Linux
      ```
    - GKE (k8s: `v1.27.8-gke.1067004`)
      - Node OS: `ContainerOS`
        - without network policy (Calico)
          - kuma-init
        - with network policy (Calico)
          - kuma-init
          - CNI
      - Node OS: `Ubuntu`
        - with network policy (Calico)
          - kuma-init
          - CNI
    - EKS (k8s: `v1.29.1-eks-b9c9ed7`)
      - kuma-init
      - CNI
    - AKS (k8s: `v1.28.5`, network plugin: `Azure CNI`)
      - kuma-init
      - CNI
    - VM RHEL 9.3 (deployed on AWS)
      ```sh
      uname -a
      # Linux ip-172-31-0-218.eu-central-1.compute.internal 5.14.0-362.18.1.el9_3.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 3 15:54:45 EST 2024 x86_64 x86_64 x86_64 GNU/Linux
      ```
      ```sh
      docker network create kind
      docker run -ti --entrypoint="" --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose --redirect-all-dns-traffic
      docker run -ti --entrypoint="" --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose
      docker run -ti --entrypoint="" --network=kind --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose --redirect-all-dns-traffic
      docker run -ti --entrypoint="" --network=kind --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose
      ```
    - VM RHEL 8.9 (deployed on AWS)
      ```sh
      uname -a
      # Linux ip-172-31-11-4.eu-central-1.compute.internal 4.18.0-513.11.1.el8_9.x86_64 #1 SMP Thu Dec 7 03:06:13 EST 2023 x86_64 x86_64 x86_64 GNU/Linux
      ```
      ```sh
      docker network create kind
      docker run -ti --entrypoint="" --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose --redirect-all-dns-traffic
      docker run -ti --entrypoint="" --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose
      docker run -ti --entrypoint="" --network=kind --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose --redirect-all-dns-traffic
      docker run -ti --entrypoint="" --network=kind --privileged kumahq/kuma-init:0.0.0-preview.vb1ab830e1 kumactl install transparent-proxy --kuma-dp-uid 5678 --verbose
      ```
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [x] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Review by michaelbeaumont — APPROVED** [review]

File: pkg/transparentproxy/iptables/builder/builder_restore.go [hunk 1]
```diff
@@ -87,79 +87,115 @@ func (e executable) exec(ctx context.Context, args ...string) (*bytes.Buffer, er
...
 
 	// Found DOCKER_OUTPUT chain in iptables-nft
 	if nft.foundDockerOutputChain {
-		return nft, false, nil
+		return nft.withFallback(legacy), nil
 	}
 
-	if err := legacy.verify(ctx, cfg); err != nil {
-		return nft, false, nil
+	if legacyVerifyErr != nil {
```
Comment (slonka) [inline pkg/transparentproxy/iptables/builder/builder_restore.go:191]: shouldn't this be above the `if nft.foundDockerOutputChain` statement? like we can set fallback to legacy and have `legacyVerifyErr` right?

**Review by slonka — COMMENTED** [review]
2 questions in code and one here: should some of the parameters be fields on the config instead?

File: pkg/transparentproxy/iptables/builder/builder.go [hunk 2]
```diff
@@ -119,84 +123,105 @@ func createRulesFile(ipv6 bool) (*os.File, error) {
...
 
-	rules, err := BuildIPTables(cfg, dnsServers, ipv6)
-	if err != nil {
-		return "", fmt.Errorf("unable to build iptable rules: %s", err)
-	}
+	maxRetries := pointer.Deref(r.cfg.Retry.MaxRetries)
 
-	if err := saveIPTablesRestoreFile(cfg.RuntimeStdout, rulesFile, rules); err != nil {
-		return "", fmt.Errorf("unable to save iptables restore file: %s", err)
+	for i := 0; i <= maxRetries; i++ {
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/transparentproxy/iptables/builder/builder.go:166]: This patch is not modifying retries logic, which was added in the past to mitigate issue with xtables lock

File: pkg/transparentproxy/iptables/builder/builder_restore.go [hunk 3]
```diff
@@ -87,79 +87,115 @@ func (e executable) exec(ctx context.Context, args ...string) (*bytes.Buffer, er
...
-	if err := nft.verify(ctx, cfg); err != nil {
-		return legacy, true, legacy.verify(ctx, cfg)
+	if nftVerifyErr != nil {
+		return legacy, nil
 	}
 
 	// Found DOCKER_OUTPUT chain in iptables-nft
 	if nft.foundDockerOutputChain {
-		return nft, false, nil
+		return nft.withFallback(legacy), nil
```
Comment (slonka) [inline pkg/transparentproxy/iptables/builder/builder_restore.go:188]: `withFallback` checks for `nil` and does not assign it if it's `nil`

File: pkg/transparentproxy/iptables/builder/builder_restore.go (hunk 1 above)
Comment (slonka) [reply pkg/transparentproxy/iptables/builder/builder_restore.go:191]: not relevant (see https://github.com/kumahq/kuma/pull/9776/files/74f853bfea5d1cde6743176f7acd15b0741ab42f#r1547940971)

**Pull request merged**